src/clients/arclib/arcsync.1.in
.\" -*- nroff -*-
.TH ARCSYNC 1 "@DATE" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arcsync \- ARC Synchronize

.SH DESCRIPTION

The \fBarcsync\fR command synchronizes your local jobs list with the
information at a given cluster or index server.

.SH SYNOPSIS

.B arcsync [options]

.SH OPTIONS

.IP "\fB-c\fR, \fB--cluster\fR=\fIname\fR"
select one or more computing elements: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-I\fR, \fB--infointerface\fR=\fIInterfaceName\fR"
the computing element specified by URL at the command line should be
queried using this information interface (possible options:
org.nordugrid.ldapng, org.nordugrid.ldapglue2,
org.nordugrid.wsrfglue2, org.ogf.glue.emies.resourceinfo)
.IP "\fB-g\fR, \fB--index\fR=\fIname\fR"
select one or more registries: \fIname\fR can be an alias for a single registry, a group of registries or a URL
.IP "\fB-j\fR, \fB--joblist\fR=\fIfilename\fR"
the file storing information about active jobs (default ~/.arc/jobs.xml)
.IP "\fB-f\fR, \fB--force\fR"
don't ask for verification
.IP "\fB-T\fR, \fB--truncate\fR"
truncate the joblist before synchronizing
.IP "\fB-P\fR, \fB--listplugins\fR"
list the available plugins
.IP "\fB-t\fR, \fB--timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-v\fR, \fB--version\fR"
print version information
.IP "\fB-?\fR, \fB--help\fR"
print help
.LP

.LP

.SH EXTENDED DESCRIPTION

The ARC user interface keeps a local list of active jobs in the users
home directory (see \fBarcsub\fR(1)). If this file is lost, or the user
wants to recreate the file on a different computer,the \fBarcsync\fR
command can be used to recreate the file from the information available
the specified cluster or index server.

Since the information about a job retrieved from a cluster can be
slightly out of date if the user very recently submitted or removed a
job a warning is issued when this command is run. The \fB--force\fR
option disables this warning.

The
.B --cluster
options can be used to select or reject clusters that should be
considered in the synchronization. This options can be repeated several
times.  See
.BR arcsub (1)
for a discussion of the format of arguments to this option.

If the joblist is not empty when invoking synchronization the old jobs
will be merged with the new jobs, unless the
.B --truncate
option is given, in which case the joblist will first be truncated and
then the new jobs will be added.

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the
ARC client configuration file. By using the
.B --conffile
option a different configuration file can be used than the default.

.TP
.B ~/.arc/jobs.xml
This is a local list of the user's active jobs. When a job is
successfully submitted it is added to this list and when it is removed
from the remote cluster it is removed from this list. This list is
used as the list of all active jobs when the user specifies the
.B --all
option to the various NorduGrid ARC user interface commands. By using the
.B --joblist
option a different file can be used than the default.

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT

APACHE LICENSE Version 2.0

.SH AUTHOR

Mattias Ellert <mattias.ellert@fysast.uu.se>

.SH SEE ALSO

.BR arccat (1),
.BR arcclean (1),
.BR arccp (1),
.BR arcget (1),
.BR arcinfo (1),
.BR arckill (1),
.BR arcls (1),
.BR arcmigrate (1),
.BR arcmkdir (1),
.BR arcproxy (1),
.BR arcrenew (1),
.BR arcresub (1),
.BR arcresume (1),
.BR arcrm (1),
.BR arcstat (1),
.BR arcsub (1),
.BR arctest (1)



NEWS
no news good news :)


src/clients/arclib/arcsync.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <list>
#include <string>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include <arc/ArcLocation.h>
#include <arc/compute/Endpoint.h>
#include <arc/DateTime.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/StringConv.h>
#include <arc/UserConfig.h>
#include <arc/compute/EntityRetriever.h>
#include <arc/compute/Job.h>

class JobSynchronizer : public Arc::EntityConsumer<Arc::Endpoint> {
public:
  JobSynchronizer(
    const Arc::UserConfig& uc,
    const std::list<Arc::Endpoint>& services,
    const std::list<std::string>& rejectedServices = std::list<std::string>(),  
    const std::set<std::string>& preferredInterfaceNames = std::set<std::string>(),
    const std::list<std::string>& capabilityFilter = std::list<std::string>(1, Arc::Endpoint::GetStringForCapability(Arc::Endpoint::COMPUTINGINFO))
  ) : uc(uc), ser(uc, Arc::EndpointQueryOptions<Arc::Endpoint>(true, capabilityFilter, rejectedServices)),
      jlr(uc, Arc::EndpointQueryOptions<Arc::Job>(preferredInterfaceNames))
  {
    ser.addConsumer(*this);
    jlr.addConsumer(jobs);
    
    for (std::list<Arc::Endpoint>::const_iterator it = services.begin(); it != services.end(); it++) {
      if (it->HasCapability(Arc::Endpoint::REGISTRY)) {
        ser.addEndpoint(*it);
      } else {
        jlr.addEndpoint(*it);
      }
    }
  }

  void wait() {
    ser.wait();
    jlr.wait();
  }

  void addEntity(const Arc::Endpoint& service) {
    if (service.HasCapability(Arc::Endpoint::COMPUTINGINFO)) {
      jlr.addEndpoint(service);
    }
  }

  bool writeJobs(bool truncate) {
    bool jobsWritten = false;
    bool jobsReported = false;
    // Write extracted job info to joblist
    if (truncate) {
      if ( (jobsWritten = Arc::Job::WriteJobsToTruncatedFile(uc.JobListFile(), jobs)) ) {
        for (std::list<Arc::Job>::const_iterator it = jobs.begin();
             it != jobs.end(); it++) {
          if (!jobsReported) {
            std::cout << Arc::IString("Found the following jobs:")<<std::endl;
            jobsReported = true;
          }
          if (!it->Name.empty()) {
            std::cout << it->Name << " (" << it->JobID << ")" << std::endl;
          }
          else {
            std::cout << it->JobID << std::endl;
          }
        }
        std::cout << Arc::IString("Total number of jobs found: ") << jobs.size() << std::endl;
      }
    }
    else {
      std::list<const Arc::Job*> newJobs;
      std::set<std::string> prunedServices;
      jlr.getServicesWithStatus(Arc::EndpointQueryingStatus::SUCCESSFUL,
                                prunedServices);
      if ( (jobsWritten = Arc::Job::WriteJobsToFile(uc.JobListFile(), jobs,
                                                    prunedServices, newJobs)) ) {
        for (std::list<const Arc::Job*>::const_iterator it = newJobs.begin();
             it != newJobs.end(); it++) {
          if (!jobsReported) {
            std::cout << Arc::IString("Found the following new jobs:")<<std::endl;
            jobsReported = true;
          }
          if (!(*it)->Name.empty()) {
            std::cout << (*it)->Name << " (" << (*it)->JobID << ")" << std::endl;
          }
          else {
            std::cout << (*it)->JobID << std::endl;
          }
        }
        std::cout << Arc::IString("Total number of new jobs found: ") << newJobs.size() << std::endl;
      }
    }

    if (!jobsWritten) {
      std::cout << Arc::IString("ERROR: Failed to lock job list file %s", uc.JobListFile()) << std::endl;
      std::cout << Arc::IString("Please try again later, or manually clean up lock file") << std::endl;
      return false;
    }

    return true;
  }

private:
  const Arc::UserConfig& uc;
  Arc::ServiceEndpointRetriever ser;
  Arc::JobListRetriever jlr;
  Arc::EntityContainer<Arc::Job> jobs;
};

#include "utils.h"

int RUNMAIN(arcsync)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::Logger logger(Arc::Logger::getRootLogger(), "arcsync");
  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  ClientOptions opt(ClientOptions::CO_SYNC, " ",
                    istring("The arcsync command synchronizes your "
                            "local job list with the information at\n"
                            "the given resources or index servers."));

  std::list<std::string> params = opt.Parse(argc, argv);

  if (opt.showversion) {
    std::cout << Arc::IString("%s version %s", "arcsync", VERSION)
              << std::endl;
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!opt.debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(opt.debug));

  if (opt.show_plugins) {
    std::list<std::string> types;
    types.push_back("HED:JobListRetrieverPlugin");
    showplugins("arcsync", types, logger);
    return 0;
  }

  Arc::UserConfig usercfg(opt.conffile, opt.joblist);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (!checkproxy(usercfg)) {
    return 1;
  }

  if (opt.debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  //sanity check
  if (!opt.forcesync) {
    std::cout << Arc::IString("Synchronizing the local list of active jobs with the information in the\n"
                              "information system can result in some inconsistencies. Very recently submitted\n"
                              "jobs might not yet be present, whereas jobs very recently scheduled for\n"
                              "deletion can still be present."
                              ) << std::endl;
    std::cout << Arc::IString("Are you sure you want to synchronize your local job list?") << " ["
              << Arc::IString("y") << "/" << Arc::IString("n") << "] ";
    std::string response;
    std::cin >> response;
    if (Arc::lower(response) != std::string(Arc::FindTrans("y"))) {
      std::cout << Arc::IString("Cancelling synchronization request") << std::endl;
      return 0;
    }
  }

  std::list<Arc::Endpoint> endpoints = getServicesFromUserConfigAndCommandLine(usercfg, opt.indexurls, opt.clusters);

  std::list<std::string> rejectDiscoveryURLs = getRejectDiscoveryURLsFromUserConfigAndCommandLine(usercfg, opt.rejectdiscovery);

  if (endpoints.empty()) {
    logger.msg(Arc::ERROR, "No services specified. Please configure default services in the client configuration,"
                           "or specify a cluster or index (-c or -g options, see arcsync -h).");
    return 1;
  }


  std::set<std::string> preferredInterfaceNames;
  if (usercfg.InfoInterface().empty()) {
    preferredInterfaceNames.insert("org.nordugrid.ldapglue2");
  } else {
    preferredInterfaceNames.insert(usercfg.InfoInterface());
  }

  JobSynchronizer js(usercfg, endpoints, rejectDiscoveryURLs, preferredInterfaceNames);
  js.wait();
  return js.writeJobs(opt.truncate)?0:1; // true -> 0, false -> 1.
}
